### PR TITLE
add windows support for root CA cert stores

### PR DESCRIPTION
**Pull request by rosskirkpat**
### Summary

Move the get/create of a root CA cert store into OS dependent functions

Fixes https://github.com/rancher/system-agent/issues/83
Unblocks https://github.com/rancher/windows/issues/161

### Occurred changes and/or fixed issues

 Windows does not support x509.SystemCertPool() in <go1.18. This PR implements a workaround to load the System Cert Store on Windows nodes.

### Technical notes summary

This is adapted from the following workaround: https://github.com/golang/go/issues/16736#issuecomment-540373689

Error seen on Windows machines:
`Message : error loading system cert pool for probe (kubelet): crypto/x509: system root pool is not available on Windows`

Since <go1.18 on Windows cannot load the root CA cert store, this PR leverages syscalls to load the certificate context from the Root CA cert store and then create a new x509 cert pool that contains the Root CAs. This will likely cause a performance drop on Windows as each probe that requires a CACert will need to build it's own x509 cert pool with the root CAs.

Once system-agent is upgraded to go1.18+ the differing code paths should no longer be required.

### Areas or cases that should be tested

This PR cannot be tested before it is merged as far as I am aware.

### Areas which could experience regressions

These changes should only impact system agent functionality on Windows. The existing functionality for linux hosts has been shuffled into a new function but the underlying code is still the same. For windows, given that loading the CA certs from the root store is currently broken I do not see any potential for regression.

### Screenshot/Video
N/A

### CODE REVIEW

**Review by phillipsj — APPROVED** [review]

File: pkg/prober/prober_windows.go
Comment (thedadams) [inline pkg/prober/prober_windows.go:55]: Why do we need to copy the certificate? I understand that `x509.ParseCertificates` doesn't create its own copy, but why is that a problem?

It also looks like you are creating another copy with `buf2`. So we seem to have 3 copies?

In addition, why do we create an array here (with `[1 << 20]byte`) instead of creating a slice?

File: pkg/prober/prober_windows.go
Comment (thedadams) [inline pkg/prober/prober_windows.go:62]: It looks like `certs` will only ever have one certificate added to it. Why do we need `certs` here?

**Review by thedadams — CHANGES_REQUESTED** [review]

File: pkg/prober/prober.go
Comment (rosskirkpat) [reply pkg/prober/prober.go:60]: added a nil check

File: pkg/prober/prober_windows.go
Comment (rosskirkpat) [reply pkg/prober/prober_windows.go]: I fixed this, I was missing the for loop that contains this code block

File: pkg/prober/prober_windows.go
Comment (brandond) [inline pkg/prober/prober_windows.go:86]: Can we perhaps just slice this to the correct length, instead of allocating a second slice of the correct length and then copying the bytes over?

```suggestion
		buf := (*[1 << 20]byte)(unsafe.Pointer(certContext.EncodedCert))[:certContext.Length]

		// validate the root CA certificate and return a x509.Certificate pointer
		// that is appended into our array of x509 certificates
		c, err := x509.ParseCertificate(buf)
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/prober/prober_windows.go
Comment (rosskirkpat) [reply pkg/prober/prober_windows.go:86]: I wasn't sure how to do this but I knew there was a better way. thank you @brandond !

File: pkg/prober/prober_windows.go
Comment (rosskirkpat) [reply pkg/prober/prober_windows.go]: This has been addressed with the latest commit.

**Review by luthermonson — APPROVED** [review]

**Comment by rosskirkpat** [comment]
@snasovich If @Oats87 is not able to review this by EOD today as he is out of office, can I get your approval to merge once I have another round of approvals from @brandond and @thedadams ?

File: pkg/prober/prober_windows.go
Comment (brandond) [inline pkg/prober/prober_windows.go:78]: One more question - you said up above that
> 		// `CertEnumCertificatesInStore` will iterate through
>		// all available certContexts in the specified certificate store and
>		// return a single certContext containing the next certificate in the store

in that case, the certContext should only ever be 16kb, the max size of a single certificate entry, right? We don't need to over-allocate for the size of the entire store, since that's not what this is going to point at.

File: pkg/prober/prober.go
Comment (thedadams) [reply pkg/prober/prober.go:60]: I don't see a `nil`-check here.

It is possible for `GetSystemCertPool` to return `nil, error` in many cases, both for `unix` and `windows`. If this happens, we are only logging the error and continuing. Which means that `caCertPool` will be `nil` as we continue lower. This may cause later calls (like the on on line 71) to panic.

File: pkg/prober/prober_unix.go
Comment (thedadams) [inline pkg/prober/prober_unix.go:23]: I think this can be removed.

**Review by thedadams — CHANGES_REQUESTED** [review]

File: pkg/prober/prober_windows.go
Comment (rosskirkpat) [reply pkg/prober/prober_windows.go:78]: I actually misinterpreted the syscall and that comment is not a proper description. It only fetches one cert at a time. I will update the comments I included to indicate that.

I am using the workaround referenced here, https://github.com/golang/go/issues/16736#issuecomment-540373689, which does not explain why they chose `(1*2)^20`. I assume it had to do with being divisible by 8 but it does make more sense to use 16kb. I'll get this change added

Comment (brandond) [reply pkg/prober/prober_windows.go:78]: Hmm, I see some similar work in https://github.com/cloudfoundry/systemcerts/blob/master/root_windows.go, and I can't find anything in the MS API docs regarding the max size for the [_CERT_CONTEXT struct](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_context), so I guess 1024k is a fine worst-case value. You might update the comment to indicate that we're sizing for a single context (certificate) and not the whole store, though. That was a little confusing for me.

File: pkg/prober/prober_unix.go
Comment (rosskirkpat) [reply pkg/prober/prober_unix.go:23]: I added the nil check inside each of the functions for linux/windows as it allows the function to fail faster for Windows and avoid a situation where a nil pointer causes a failure before we could check it in the main prober.go. Waiting to check until after the caCertPool is returned could result in a nil pointer being returned when we try to append the certs from the syscall to the caCertPool.

File: pkg/prober/prober_windows.go
Comment (brandond) [reply pkg/prober/prober_windows.go:80]: Since we can't figure out exactly how large it'll be, maybe lets guard against making it too small? Otherwise, if we error on the short size, we'll crash when trying to access past the end of the array.

```suggestion
		// put this up at the top
		const maxEncodedCertLen := 1 << 20
		if certContext.Length > maxEncodedCertLen {
				return fmt.Errorf("invalid CertContext length %d, certContext.Length)
		}
		buf := (*[maxEncodedCertLen]byte)(unsafe.Pointer(certContext.EncodedCert))[:certContext.Length]
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/prober/prober_windows.go
Comment (rosskirkpat) [reply pkg/prober/prober_windows.go]: agree w/ guarding against that situation, pushed the changes you suggested

**Review by brandond — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**
